### PR TITLE
Adds @defer to the GraphQL core.

### PR DESCRIPTION
**Pull request by michaelstaib**

### CODE REVIEW

File: src/HotChocolate/Core/src/Abstractions/Execution/IQueryResultBuilder.cs [hunk 1]
```diff
@@ -27,6 +27,12 @@ public interface IQueryResultBuilder
 
         IQueryResultBuilder ClearContextData();
 
+        IQueryResultBuilder SetLabel(string? label);
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Abstractions/Execution/IQueryResultBuilder.cs:30]: documentation

File: src/HotChocolate/Core/src/Execution/Processing/SelectionSetExtensions.cs [hunk 2]
```diff
@@ -29,11 +29,31 @@ internal static class SelectionSetExtensions
...
                             parent,
-                            createPath(selection.ResponseName),
+                            path.Append(selection.ResponseName),
                             scopedContext));
                 }
             }
 
+            if (selectionSet.Fragments.Count > 0)
+            {
+
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Execution/Processing/SelectionSetExtensions.cs:39]: Line

File: src/HotChocolate/Core/src/Types/Types/Directives/DeferDirectiveType.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    public class DeferDirectiveType
+        : DirectiveType<DeferDirective>
+    {
+        protected override void Configure(
+            IDirectiveTypeDescriptor<DeferDirective> descriptor)
+        {
+            descriptor
+                .Name("defer")
+                .Description(
+                    "The `@defer` directive may be provided for fragment spreads and " +
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types/Types/Directives/DeferDirectiveType.cs:24]: Resouces

File: src/HotChocolate/Core/src/Types/Types/Directives/StreamDirectiveType.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    public class StreamDirectiveType
+        : DirectiveType
+    {
+        protected override void Configure(
+            IDirectiveTypeDescriptor descriptor)
+        {
+            descriptor
+                .Name("stream")
+                .Description(
+                    "The `@stream` directive may be provided for a field of `List` " +
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types/Types/Directives/StreamDirectiveType.cs:22]: Resources

File: src/HotChocolate/Core/test/Types.Tests/DataLoader/Types/SubscriptionTypeTests.cs [hunk 5]
```diff
@@ -246,11 +246,11 @@ public async Task SubscribeAndResolve_Attribute_AsyncEnumerable(string field)
         }
 
         [InlineData("onSomething")]
-        [InlineData("onSomethingTask")]
-        [InlineData("onSomethingValueTask")]
-        [InlineData("onSomethingObj")]
-        [InlineData("onSomethingObjTask")]
-        [InlineData("onSomethingObjValueTask")]
+        // [InlineData("onSomethingTask")]
```
Comment (PascalSenn) [inline src/HotChocolate/Core/test/Types.Tests/DataLoader/Types/SubscriptionTypeTests.cs:249]: lets not forget these

File: src/HotChocolate/Core/src/Abstractions/RootPathSegment.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        public override NamePathSegment Append(NameString name) =>
+            New(name);
+
+        /// <inheritdoc />
+        public override string Print() => "/";
+
+        /// <inheritdoc />
+        public override bool Equals(Path? other)
+        {
+            if (ReferenceEquals(other, null))
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Abstractions/RootPathSegment.cs:39]: is null ?

File: src/HotChocolate/Core/src/Execution/Pipeline/OperationExecutionMiddleware.cs [hunk 7]
```diff
@@ -103,16 +105,41 @@ internal sealed class OperationExecutionMiddleware
...
                                 context.Variables);
 
-                            context.Result = await _mutationExecutor
+                            result = await _mutationExecutor
                                 .ExecuteAsync(operationContext)
                                 .ConfigureAwait(false);
                         }
 
+                        if (operationContext.Execution.DeferredTaskBacklog.IsEmpty || 
+                            result is null)
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Execution/Pipeline/OperationExecutionMiddleware.cs:114]: can the inital result of defere be null?

File: src/HotChocolate/Core/test/Types.Tests/DataLoader/Types/SubscriptionTypeTests.cs (hunk 5 above)
Comment (michaelstaib) [reply src/HotChocolate/Core/test/Types.Tests/DataLoader/Types/SubscriptionTypeTests.cs:249]: good catch haha

File: src/HotChocolate/Core/src/Abstractions/RootPathSegment.cs (hunk 6 above)
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Abstractions/RootPathSegment.cs:39]: that is correct

File: src/HotChocolate/Core/src/Execution/Pipeline/OperationExecutionMiddleware.cs (hunk 7 above)
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Execution/Pipeline/OperationExecutionMiddleware.cs:114]: theoretically, if a new operation is introduced, in this case the result will be null and we yield to the next middleware.

File: src/HotChocolate/Core/test/Types.Tests/DataLoader/Types/SubscriptionTypeTests.cs (hunk 5 above)
Comment (michaelstaib) [reply src/HotChocolate/Core/test/Types.Tests/DataLoader/Types/SubscriptionTypeTests.cs:249]: good catch

**Pull request merged**